gammapy/detect/cwt.py

```diff
@@ -188,127 +197,159 @@ def set_data(self, counts, background):
+              of 'scipy.ndimage.label')
+            - for each structure do:
+                - find pixels coordinates of the structure
+                - if just one pixel, we can remove it from mask if we want
+                - if the max value of significance in that structure less
+                  than nsigmap, remove this structure from mask
+            - update support by mask
         """
         from scipy.ndimage import label
         # TODO: check that transform has been performed
 
-        sig = self.transform / self.error
+        log.info('Computing significance')
+        significance = self.transform / self.error
+
+        log.info('For each scale start to compute support')
+        for idx_scale in range(self.n_scale):
+            log.info('Start to compute support for scale {:.2f}'.format(idx_scale))
 
-        for key in self.scales.keys():
-            tmp = sig[key] > nsigma
-            # produce a list of connex structures in the support
-            l, n = label(tmp)
-            for id in range(1, n):
-                index = np.where(l == id)
-                if remove_isolated:
-                    if index[0].size == 1:
-                        tmp[index] *= 0.0  # Remove isolated pixels from support
-                signif = sig[key][index]
-                if signif.max() < nsigmap:  # Remove significant pixels island from support
-                    tmp[index] *= 0.0  # if max does not reach maximal significance
+            log.info('Create mask based on lower threshold nsigma {:.2f}'.format(self.nsigma))
+            mask = significance[idx_scale] > self.nsigma
 
-            self.support[key] += tmp
-            self.support[key] = self.support[key] > 0.
+            # Produce a list of connex structures in the support
+            labeled_mask, n_structures = label(mask)
+            for struct_label in range(1, n_structures + 1):
+                coords = np.where(labeled_mask == struct_label)
 
-        return self.support
+                # Remove isolated pixels from support
+                if self.remove_isolated and coords[0].size == 1:
+                    log.info('Remove isolated pixels from support')
+                    mask[coords] = False
 
-    def inverse_transform(self):
+                # If maximal significance of the structure does not reach nsigmap(maximal
+                # significance), remove significant pixels island from support
+                struct_signif = significance[idx_scale][coords]
+                if struct_signif.max() < self.nsigmap:
+                    log.info('Remove significant pixels island {0} from support'.format(struct_label))
+                    mask[coords] = False
+
+            log.info('Update supprort for scale {:.2f}'.format(idx_scale))
+            self.support[idx_scale] += mask
+            self.support[idx_scale] = self.support[idx_scale] > 0.
+
+    def _inverse_transform(self):
```

Comment:
@cdeil @adonath I don't understand what's going on here. I try to understand tomorrow but maybe you have thoughts about it.